src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -57,7 +57,27 @@ private void ReturnReadBufferIfEmpty()
 
         ~SslStreamInternal()
         {
-            if (_internalBuffer != null)
+            Dispose(disposing: false);
+        }
+
+        public void Dispose()
+        {
+            Dispose(disposing: true);
+
+            if (_internalBuffer == null)
+            {
+                // Suppress finalizer if the read buffer was returned.
+                GC.SuppressFinalize(this);
+            }
+        }
+
+        private void Dispose(bool disposing)
+        {
+            if (_internalBuffer != null &&
```

Comment:
If there is a concurrent operation, is it possible the buffer could go from non-null to null and then we'd hit an exception trying to return null to the pool? It should probably first be read into a local and then that local used subsequently. Also, doesn't a read completing sometimes return a buffer, or am I mistaken? If it does, couldn't we end up double-returning?

Reply:
Seems like to make this safe it needs to instead be something like:
```C#
private void Dispose(bool disposing)
{
    if (Interlocked.Exchange(ref _nestedRead, 1) == 0)
    {
        byte[] buffer = _internalBuffer;
        if (buffer != null)
        {
            _internalBuffer = null;
            ArrayPool<byte>.Shared.Return(buffer);
        }
    }
}
```
so that only check the buffer after we've already confirmed that there's no active read operation and ensured no additional one can be started (that's what _nestedRead does, right?).

Reply:
(We may also want to rename _internalBuffer to _readBuffer, or something like that to clarify that it's for reading, not writing; there's a comment on it to this effect, but it's not clear from the name.  If we do rename it, we'll obviously want to rename the related fields as well.)

Reply:
A completing read returns the buffer only if it is empty, and by returning it, it sets the _internalBuffer to null. Also, when a read operation is in progress, this will just return, so this piece of code will not run concurrent to an ongoing read operation.

Reply:
@Priya91, I don't believe that's true.  For example, consider this interleaving:
1. Thread 1 calls Dispose.  It gets as far as checking that _internalBuffer != null and see that's true.
2. Thread 2 is in the middle of doing its read.  It sets _internalBuffer to null and exits the read.
3. Thread 1 now continues on, sees that there's no current read, and returns null to the ArrayPool.  Kaboom.

Reply:
Hmm yeah, so the Read check should come first and then the buffer check

Reply:
Also, when the finalizer is being run, isn't it guaranteed that no other Read operation will be ongoing, since the finalizer is being run only when the object is marked for deallocation by GC.. So _nestedRead check is not required at the time of finalization?

Reply:
Probably, but I don't think it's worth here trying to optimize away an interlocked that occurs during finalization, especially since with your change we hope that finalization will be very rare, and if it does happen, you're already incurring significant other costs.

Reply:
Separately... does this change impact anything about behavior after SslStream.Dispose is called?  This is causing nestedRead to be set to 1 on Dispose; will a subsequent Read{Async} see that and throw an exception about their being a concurrent read operation?  Or do such operations first check for disposal and so it wouldn't get that far?

Reply:
Yeah they check for disposal, so subsequent Read operations throw ObjectDisposedException, I've added this check in the new test in this PR..